enshrine-esdt-safe/src/from_sovereign/transfer_tokens.rs

```diff
@@ -36,13 +39,66 @@ pub trait TransferTokensModule:
             sc_panic!("Operation is not registered");
         }
 
+        let are_tokens_registered =
+            self.verify_operation_tokens_issue_paid(operation.tokens.clone());
+
+        if !are_tokens_registered {
+            self.emit_transfer_failed_events(
+                &hash_of_hashes,
+                &OperationTuple {
+                    op_hash: operation_hash.clone(),
+                    operation: operation.clone(),
+                },
+            );
+        }
+
         let minted_operation_tokens = self.mint_tokens(&operation.tokens);
         let operation_tuple = OperationTuple {
-            op_hash: operation_hash,
-            operation,
+            op_hash: operation_hash.clone(),
+            operation: operation.clone(),
         };
 
-        self.distribute_payments(hash_of_hashes, operation_tuple, minted_operation_tokens);
+        self.distribute_payments(
+            hash_of_hashes.clone(),
+            operation_tuple,
+            minted_operation_tokens,
+        );
+    }
+
+    #[endpoint(registerTokens)]
+    #[payable("*")]
+    fn register_tokens(&self, tokens: MultiValueEncoded<TokenIdentifier>) {
+        let call_payment = self.call_value().single_esdt().clone();
+        let wegld_identifier = self.wegld_identifier().get();
+
+        require!(
+            call_payment.token_identifier == wegld_identifier,
+            "WEGLD is the only token accepted as register fee"
+        );
+
+        require!(
+            call_payment.amount == DEFAULT_ISSUE_COST * tokens.len() as u64,
+            "WEGLD fee amount is not met"
+        );
+
+        for token_id in tokens {
+            self.register_token(token_id);
+        }
+    }
+
+    fn verify_operation_tokens_issue_paid(
+        &self,
+        tokens: ManagedVec<OperationEsdtPayment<Self::Api>>,
+    ) -> bool {
+        require!(!tokens.is_empty(), "Tokens array should not be empty");
```

Comment:
You said MultiESDTNFTTransfer cannot have 0 tokens as payment so this require is not needed. Also this require should never happen because will fail the transactions and not send the fail events